게시판 Redis 적용

K-MG-0328 opened the pull request:
[feat(board): 조회수 증가 기능에 Redis 캐싱 추가](https://github.com/K-MG-0328/BigBoardRestAPI/commit/e5a44c91f261a7286570fded73ecbb1c260359b3) 

- Redis를 사용해 조회수 데이터를 캐싱하도록 구현
- 키에 TTL 1일 설정 추가
- 매 분마다 데이터베이스와 동기화 (스케줄링)

feat(board): 게시판 목록 기능에 Redis 캐싱 추가
- Redis를 사용해 게시판 목록을 캐싱하도록 구현
- 키에 TTL 5분 설정 추가
- 매 분마다 데이터베이스와 동기화 (스케줄링)

fix(Entity): 조회수 캐싱 시 Redis 키 포맷 오류 수정
- Redis 키 생성 시 잘못된 형식 문제 해결
- LocalDateTime Redis 캐시 저장시 직렬화 오류로 인해 발생한 InvalidDefinitionException 오류 수정

docs(redis): Redis 설정 관련 주석 추가
- 캐싱 정책 및 데이터 구조 설명 포함

[feat : 조회수 캐싱 동시성 처리, 스케줄링 변경](https://github.com/K-MG-0328/BigBoardRestAPI/commit/f5e16c5afe5fe7d0580f9ee2381eefe9418b4139) 

- RedisViewCountService : 동시성 문제 방지를 위한 동기화 코드 추가.
- ViewCountSyncScheduler : DB 동기화 스케줄링 10초 -> 60초로 변경

refactor : 로직 분리
- Redis 적용 전, 적용 후 컨트롤러, 서비스, 리포지토리 분리
- RedisKeyHelper uitl 패키지로 이동


[feat : 평점 캐싱 기능](https://github.com/K-MG-0328/BigBoardRestAPI/commit/95855604cb2b76ae542533046c7ffe723d638c6c) 

- 평점 적용시 캐싱 -> 스케줄링을 통한 DB에 일괄 적용

refactor : 로직 분리
- Redis 적용 전, 적용 후 컨트롤러, 서비스, 리포지토리 분리

K-MG-0328 committed:
코드 최적화

K-MG-0328 committed:
게시판 평점 구현 및 DTO 클래스 생성

K-MG-0328 committed:
게시판 평점 구현 및 DTO 클래스 생성

K-MG-0328 committed:
viewCount, totalScore,ratingCount 원시타입으로 변경

K-MG-0328 committed:
오류 수정

K-MG-0328 committed:
기능 테스트 후 수정

K-MG-0328 committed:
설정정보

K-MG-0328 committed:
동기화

K-MG-0328 committed:
Remove .gitignore from tracking

K-MG-0328 committed:
설정파일

K-MG-0328 committed:
엔티티 수정

K-MG-0328 committed:
Redis 설정

K-MG-0328 committed:
Redis 적용

K-MG-0328 committed:
feat(board): 조회수 증가 기능에 Redis 캐싱 추가
- Redis를 사용해 조회수 데이터를 캐싱하도록 구현
- 키에 TTL 1일 설정 추가
- 매 분마다 데이터베이스와 동기화 (스케줄링)

feat(board): 게시판 목록 기능에 Redis 캐싱 추가
- Redis를 사용해 게시판 목록을 캐싱하도록 구현
- 키에 TTL 5분 설정 추가
- 매 분마다 데이터베이스와 동기화 (스케줄링)

fix(Entity): 조회수 캐싱 시 Redis 키 포맷 오류 수정
- Redis 키 생성 시 잘못된 형식 문제 해결
- LocalDateTime Redis 캐시 저장시 직렬화 오류로 인해 발생한 InvalidDefinitionException 오류 수정

docs(redis): Redis 설정 관련 주석 추가
- 캐싱 정책 및 데이터 구조 설명 포함

K-MG-0328 committed:
feat : 조회수 캐싱 동시성 처리, 스케줄링 변경
RedisViewCountService : 동시성 문제 방지를 위한 동기화 코드 추가.
ViewCountSyncScheduler : DB 동기화 스케줄링 10초 -> 60초로 변경

refactor : 로직 분리
Redis 적용 전, 적용 후 컨트롤러, 서비스, 리포지토리 분리
RedisKeyHelper uitl 패키지로 이동

K-MG-0328 committed:
feat : 평점 캐싱 기능
평점 적용시 캐싱 -> 스케줄링을 통한 DB에 일괄 적용

refactor : 로직 분리
Redis 적용 전, 적용 후 컨트롤러, 서비스, 리포지토리 분리

K-MG-0328 committed:
feat(scouter): scouter 설정 추가

test(성능 테스트):
성능 테스트를 위한  Deprecated 제거
nGrinder에서 POST 요청만 지원하므로 RequestMapping으로 변경

repactor(boardService, Controllers):
@Cacheable 어노테이션 제거
조회수 증가 로직 분리하여 Transaction 적용
Mapping URL 직관적으로 구분되게 변경

fix(boardService):
평점 추가시 Transaction 적용

K-MG-0328 committed:
Create README.md

K-MG-0328 committed:
Update README.md

K-MG-0328 committed:
Update README.md

K-MG-0328 committed:
Update README.md

f-lab-k commented on build.gradle:
```diff
@@ -0,0 +1,41 @@
+dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-data-redis'
```
버전을 명시하는 것이 좋겠습니다.
어떤 버전을 사용하는지 바로 확인이 불가능합니다
또한 서버에서 gradle 버전이 올라가면 해당 앱의 버전이 자동으로 변경 될 가능성도 있어보이고요

K-MG-0328 replied:
넵 근데 플러그인에 버전을 명시해놓으면 플러그인에 버전을 명시해놓으면 스프링부트 버전이 자동으로 맞춰지지않나요?
plugins {
    id 'java'
    id 'org.springframework.boot' version '3.4.1'
    id 'io.spring.dependency-management' version '1.1.7'
}

f-lab-k replied:
> 넵 근데 플러그인에 버전을 명시해놓으면 플러그인에 버전을 명시해놓으면 스프링부트 버전이 자동으로 맞춰지지않나요? plugins { id 'java' id 'org.springframework.boot' version '3.4.1' id 'io.spring.dependency-management' version '1.1.7' }

아 네 맞습니다
그걸 제가 놓쳤네요!

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/controller/BoardController.java:
```diff
@@ -1,43 +1,58 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/before/board")
```
@RequestMapping("/before/board")
이렇게 uri가 클래스와 메서드로 나뉘어져있으면 찾기가 어려울 것 같습니다
debugging 순서 -> 화면에서 어떤 api를 사용하는지 확인한다
그 uri를 복사해서 intellij에서 찾는다
이런 과정일 때 바로 찾기 어려워요
더군다나 프로젝트의 규모가 커지면 더 찾기 힘듭니다.

K-MG-0328 replied:
넵 수정하겠습니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/controller/RedisBoardController.java:
```diff
@@ -0,0 +1,61 @@
+
+    //게시글 생성
+    @PostMapping
+    public ResponseEntity<BoardDetailResponse> createBoard(@RequestBody BoardDetailRequest board) {
```
BoardDetailRequest가 그대로 service layer로 들어가는 것은 좋아보이지 않습니다
상위 레이어에서는 하위 레이어의 데이터를 참조할 수 있지만
하위 레이어가 상위 레이어의 데이터를 참조하는 것은 좋아보이지 않습니다
단방향으로만 참조해야 합니다
나중에 모듈로 분리할 때 분리하기 어려워져요

K-MG-0328 replied:
아 그렇군요 단방향으로 참조해야하는거군요! 감사합니다. 수정하겠습니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/controller/RedisBoardController.java:
```diff
@@ -0,0 +1,61 @@
+    //게시글 삭제
+    @DeleteMapping("/{boardId}")
+    public ResponseEntity<Void> deletBoard(@PathVariable Long boardId) {
+        boardService.deleteBoard(boardId);
```
delete, update 할 때 boardId로만 조작이 가능하다면 다른 유저가 쓴 글을 그 외 유저가 수정할 수 있게 되는데요
그렇게 되면 안되지 않을까요?

K-MG-0328 replied:
넵 userId와 authorId를 검증하는 로직 추가했습니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/controller/RedisBoardController.java:
```diff
@@ -0,0 +1,61 @@
+
+    //게시글 목록 조회 Redis 적용 후
+    @GetMapping()
+    public List<BoardResponse> getBoards(
```
모든 것에 그러한 것은 아니지만 List로 응답할 경우 스펙의 추가가 있다면 확장하기 어렵습니다
예를들어 보드 게시물 응답을 해야 할 때 해당 게시물의 총 개수나 그 외 무엇인가? 추가 된다면 확장 하기 어렵습니다
버저닝을 해야 될 거에요 그럼 그때 앱, 웹 다른 server에서도 사용한다면 다 변경해달라고 해야합니다.

K-MG-0328 replied:
아,, 그렇군요 반환을 객체로 반환하고 객체 안에 실제 데이터와 메타 데이터를 함께 반환하는 구조로 만들어야하는군요! 감사합니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/entity/Board.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public static Board toBoard(BoardDetailRequest boardDetailRequest) {
+        return Board.builder()
```
dto가 여기까지 침범을ㅎㅎ

K-MG-0328 replied:
넵 수정했습니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/service/BoardService.java:
```diff
@@ -1,59 +1,75 @@
+
+    //게시글 수정
+    public BoardDetailResponse updateBoard(BoardDetailRequest updateBoard){
+        Board board = boardRepository.findById(updateBoard.getBoardId()).orElse(null);
```
orElse를 null로 하게 될 경우 boardId가 db에 없으면 npe 발생하지 않아요?

K-MG-0328 replied:
넵 예외 발생시 의미있는 예외를 던지도록 수정했습니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/service/RedisViewCountService.java:
```diff
@@ -0,0 +1,48 @@
+        String key = "board:" + boardId + ":viewCount"; // 키 생성
+
+        /* TTL로 레디스에서 캐시가 사라지게 되면 조회수가 초기화될 가능성이 있으므로 검증로직을 추가*/
+        synchronized (this) { // 동시성 문제를 방지하기 위해 동기화 블록 추가
```
synchronized가 있어서 동시성은 지킬 수 있지만
성능은 떨어질 것 같은데요
제거 할 수 있는 방법이 있을 것 같은데요

    if (!redisTemplate.hasKey(key)) {
        int viewCount = boardRepository.getViewCount(boardId);
        
        // key가 없을 때만 값 세팅 (setIfAbsent)
        redisTemplate.opsForValue().setIfAbsent(key, String.valueOf(viewCount), Duration.ofDays(1L)); // TTL 설정
    }
    
    redisTemplate.opsForValue().increment(key); // 키의 viewCount 값 증가

그리고 ttl과 db로 싱크하는 주기가 엇갈린다면 데이터 유실이 발생할 수 있겠습니다
즉 ttl이 1분인데 만약 db 싱크하는 주기가 1분이라면 엇갈려서 반영된 조회수가 디비에 반영되지 않을 수 있어요
이것을 방지 하려면 이벤트 소싱 방법을 도입하거나
expire를 increment할 때 마다 expire를 증가시키거나 하는 방법이 있을 것 같아요

K-MG-0328 replied:
ttl과 db싱크 주기로 인한 문제 발생 여부에 대해서는 피드백 해주신 expire를 통해서 수정했습니다. 또 스레드가 동시에 뷰 카운트를 증가시키려 할 때, 자바 단에서 synchronized 로 락을 거는 대신 Redis의 원자적 연산인 setIfAbsent 를 사용해서 키가 없을 때만 DB 조회 후 레디스에 캐시 저장하도록 수정했습니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/service/RedisBoardService.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    //게시글 조회 Redis 적용 후
+    public BoardDetailResponse getBoardById(Long boardId) {
```
실시간으로 조회수를 반영하기 위하여 incrementViewCount에서 조회수 상승 시킨 값을 받고
그것을 board에 넣어주는 것이 좋지 않을까 싶어요

그리고 여기서 고민이 한가지 더 필요할 것 같아요
조회수를 비동기로 할지 말지?

조회수 증가시키는 로직이 지연되면 해당 비즈니스 로직도 영향을 받을 것 같아요
조회수 증가시키는 로직은 비동기로 하고

실시간으로 레디스에 있는 데이터를 가져오고 싶다면 get해서 가져와서 셋팅하고 응답으로 내리면 되지 않을까 싶어요

K-MG-0328 replied:
incrementViewCount는 비동기처리를 하였고 getViewCount 메소드에서 받아온 조회수 값을 board 값에 저장해서 반환하도록 수정했습니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/service/RedisRatingDataService.java:
```diff
@@ -0,0 +1,71 @@
+    private final RedisBoardRepository boardRepository;
+
+    public Set<String> getAllKeys(String pattern) {
+        Set<String> keys = redisTemplate.keys(pattern); //특정 패턴에 매칭되는 키 검색
```
keys로 가져오는 것 보다는 scan을 통해서 가져오는 것이 더 좋을 듯 합니다
keys는 사용을 지양하는 명령어 입니다.

scan을 사용해서 가져오던가 아니면 db에서 읽어와서 key를 찾는 것이 나을 것 같습니다

K-MG-0328 replied:
넵 감사합니다.

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/service/RedisRatingDataService.java:
```diff
@@ -0,0 +1,71 @@
+        return keys != null ? keys : Set.of(); // keys가 null이면 빈 Set 반환
+    }
+
+    public void updateBoardRating(BoardScore boardScore) {
```
평점 같은 경우에는 디비에 저장을 하고
주기적으로 그 평점을 계산해서 레디스에 동기화하고
게시물 요청시 레디스에 저장된 것을 함게 반영해주는 것이 나아보입니다

K-MG-0328 replied:
네 평점 db에 저장되도록 수정했고 실시간으로 평점 적용되도록 수정했습니다.

K-MG-0328 committed:
repactor(board):
BoardDetailRequest -> BoardDetailServiceRequest로 변경
BoardDetailRequest -> BoardDetailServiceRequest -> Board (상위 -> 하위) 단방향으로 참조되게 변경

add : BoardDetailServiceRequest, BoardScoreServiceRequest
service dto 생성

K-MG-0328 committed:
repactor(board):
메서드에 url 매핑

K-MG-0328 committed:
repactor(board):
board 엔티티에서 dto를 참조하지 않도록 변경
dto에서 dto -> 엔티티 변경 메소드 추가

K-MG-0328 committed:
repactor(service):
예외 발생시 의미있는 예외를 던지도록 수정

K-MG-0328 committed:
repactor(BoardListResponse):
래퍼 클래스를 반환하도록 수정

K-MG-0328 committed:
Merge remote-tracking branch 'origin/main'

f-lab-k commented on src/main/java/com/github/mingyu/bigboard/controller/BoardController.java:
```diff
@@ -1,43 +1,58 @@
+    }
+
+    //게시글 상세 조회 Redis 적용 전
+    @GetMapping("/before/board/detail/{boardId}")
```
1. 레디스 적용 전이라 그냥 before 붙이신건가요??
2. 여기에서는 detail이 있을 필요가 있을까 싶긴하네요

K-MG-0328 replied:
아 detail 제거해도 될 것 같아서 제거 했습니다. 레디스 적용 후 url하고 동일해져서 붙여놨습니다.

K-MG-0328 committed:
repactor(Redis): synchronized 제거
setIfAbsent, increment 메소드가 원자적으로 연산을 하므로 동시성에 안전함으로 불필요한 synchronized 제거

K-MG-0328 committed:
repactor(Redis): 조회수 증가 비동기 처리
조회에서 조회수 캐싱하는 로직을 비동기적으로 처리되게 수정
실시간으로 조회수를 표출하기위해 board에 조회수 값 저장해서 반환하도록 수정

repactor(상세 조회): 게시글 상세 조회
매핑 주소 url detail 제거해서 간략하게 표현

K-MG-0328 committed:
repactor(getAllKeys): 키 조회 메서드
평점 업데이트 레디스에서 Key를 가져올 때 keys 메소드 대신 scan을 통해서 가져오도록 수정
대규모 게시판 상황에서는 모든 key를 한번에 가져오게되면 풀스캔이 일어나 장애로 이어질 수 있으므로 분할로 가져오도록 수정.

K-MG-0328 committed:
feat(exception) : 게시글 수정, 삭제
게시글 수정 삭제 시 userId와 authorId를 비교하는 로직 추가 및 맞지 않을 경우 예외 처리

K-MG-0328 committed:
feat(Rating) : 평점 DB 저장되도록 수정
일괄 DB 동기화 스케줄링 주석처리
실시간으로 평점 적용되도록 수정.

K-MG-0328 committed:
fix(Caching) : ClassCastException 수정
클래스 타입 정보를 JSON에 포함되도록 설정

Refact(url) 게시글 수정, 삭제 userId 쿼리파라미터로 값을 받게 수정

K-MG-0328 committed:
fix(Caching) : RedisCommandExecutionException: WRONGTYPE Operation
키의 값이 초기에 String으로 주고 후에 해당 키에 Hash와 타입으로 값을 넣어서 생기는 문제 수정
